dds_web/api/project.py

```diff
@@ -298,3 +300,76 @@ def put(self):
                 break
 
         return flask.jsonify({"updated": updated, "error": error, "tries": current_try})
+
+
+class CreateProject(flask_restful.Resource):
+    @auth.login_required(role="admin")
+    def post(self):
+        """Create a new project"""
+        p_info = flask.request.json
+        cur_user = auth.current_user()
+        # Add check for user permissions
+
+        created_time = dds_web.utils.current_time()
+
+        try:
+            # lock Unit row
+            unit_row = (
+                db.session.query(models.Unit)
+                .filter_by(id=cur_user.unit_id)
+                .with_for_update()
+                .one_or_none()
+            )
+
+            if not unit_row:
+                raise Exception(
+                    "No unit associated to this user --> cannot create project. This message and exception should and will be changed."
+                )
+
+            unit_row.counter = unit_row.counter + 1 if unit_row.counter else 1
+            public_id = "{}{:03d}".format(unit_row.internal_ref, unit_row.counter)
+
+            project_info = {
+                "public_id": public_id,
+                "title": p_info["title"],
+                "date_created": created_time,
+                "date_updated": created_time,
+                "status": "Ongoing",  # ?
+                "description": p_info["description"],
+                "pi": p_info.get("pi", ""),  # Not a foreign key, only a name
+                "size": 0,
+                "bucket": self.__create_bucket_name(public_id, created_time),
+            }
+            pkg = key_gen.project_keygen(project_info["public_id"])
+            project_info.update(pkg.get_key_info_dict())
+
+            new_project = models.Project(**project_info)
+            unit_row.projects.append(new_project)
+            cur_user.created_projects.append(new_project)
+
+            db.session.commit()
+
+        except Exception as err:
```

Comment:
Switch this to `sqlalchemy.exc.SQLAlchemyError` instead

Reply:
Done!